src/coord/timestamp.rs

```diff
@@ -331,19 +365,33 @@ impl Timestamper {
     }
 
     /// Implements the byo timestamping logic
+    /// A new timestamp should be
+    /// 1) strictly greater than the last timestamp in this partition
+    /// 2) greater or equal to all the timestamps that have been assigned so far across all partitions
     fn update_byo_timestamp(&mut self) {
         for (id, byo_consumer) in &mut self.byo_sources {
             // Get the next set of messages from the Consistency topic
             let messages = byo_query_source(byo_consumer, self.max_increment_size);
             // Notify coordinator of updates
-            for (timestamp, offset) in byo_extract_ts_update(byo_consumer, messages) {
-                self.tx
-                    .unbounded_send(coord::Message::AdvanceSourceTimestamp {
-                        id: *id,
-                        timestamp,
-                        offset,
-                    })
-                    .expect("Failed to send update to coordinator");
+            for (partition, timestamp, offset) in byo_extract_ts_update(byo_consumer, messages) {
```

Comment:
In this section of the code, how are we sure that we're comparing partitions in the "right" order? For example, what if I had a scenario like:

Partition A gets a record at timestamp X.
Partition B gets a record at timestamp Y (>X).

Can we be sure that as we loop through `byo_extract_ts_update` that we will see Partition A before Partition B?

Reply:
When we process updates from the timestamp topic, we ensure that conditions on line 368-370 are satisfied. This is the order in which they get inserted in the vector and so the order in which they get iterated on. This is also the order in which they get sent (and therefore inserted into the worker hashmap)